### PR TITLE
feat(n4s): add enforce.tuple() for fixed-length array validation

### PR DESCRIPTION
**Pull request by ealush**
## Summary

- Add `enforce.tuple()` schema rule for validating fixed-length arrays where each position has a distinct type
- Support optional trailing elements via `enforce.optional()`, nested schemas (shape, tuple, isArrayOf), and error path propagation with index-based paths
- Add `enforce.infer<typeof schema>` type utility as an alias for `typeof schema.infer`
- Add `enforce.list()` as a semantic alias for `enforce.isArrayOf()` — identical runtime and types, shorter name that reads better alongside `shape`, `tuple`, `record`, `loose`
- Full eager/lazy API integration with correct TypeScript type inference for `.infer`, `.parse()`, and `.test()`
- Parse propagation: tuple correctly transforms values through parsers (custom coercions, built-in chains) both standalone and nested inside shape

## Test plan

- [x] 26 unit tests for tuple: basic validation, optional trailing, nested schemas, eager API, error reporting, RuleInstance methods, .message() override, and type inference
- [x] 4 parse integration tests proving custom coercions, built-in parser chains, nested shape parsing, and tuple-inside-shape transformations
- [x] 2 type inference tests for `.parse()` return types (standalone and nested in shape)
- [x] 5 tests for `enforce.list()`: validation, multiple rules, chaining, type inference, eager API
- [x] `enforce.infer<>` alias test added to integration type test suite
- [x] Pre-commit hooks pass (tsc, eslint, prettier)

## Summary by CodeRabbit

* **New Features**
  * Added `enforce.tuple()` for validating fixed-length tuples with specific type requirements for each element.
  * Added `enforce.list()` as an alternative name for array validation.
  * Added `enforce.infer<>` type utility to extract inferred types from validation rules.

* **Documentation**
  * Updated schema validation documentation with tuple examples, including optional elements and nested schemas.

### CODE REVIEW

File: packages/n4s/src/lazy.ts [hunk 1]
```diff
@@ -97,6 +97,13 @@ const schemaRulesWithArrayChaining = {
   loose: schemaAttacher(schemaEvaluators.loose),
   record: recordEvaluators.record,
   shape: schemaAttacher(schemaEvaluators.shape),
+  tuple: (...rules: any[]) =>
+    addToChain({}, (value: any) => {
```
[the remark this answers is not in the document]
Comment (ealush) [reply packages/n4s/src/lazy.ts]: Fixed in d93ce0ce — changed to `addToChain(arrayRules, ...)` for consistency with `isArrayOf`.

File: packages/n4s/src/rules/schemaRules/tuple.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+function runElementRule(
+  item: any,
+  rule: any,
+  index: number,
+): RuleRunReturn<any> {
+  return ctx.run({ value: item, set: true, meta: { index } }, () =>
+    rule.run(item),
+  );
```
[the remark this answers is not in the document]
Comment (ealush) [reply packages/n4s/src/rules/schemaRules/tuple.ts:105]: Won't fix — `tuple()` receives typed `RuleInstance` arguments via the lazy and eager APIs, so a non-rule reaching this point isn't a realistic scenario. Adding runtime guards for type-enforced invariants adds unnecessary complexity.

File: packages/n4s/src/rules/schemaRules/tuple.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+/**
+ * Checks whether a rule is an optional rule by testing if it passes
+ * with undefined (the same way shape detects optional fields).
+ */
+function isOptionalRule(rule: RuleInstance<any, any>): boolean {
+  if (!rule || typeof rule.test !== 'function') return false;
+  return rule.test(undefined);
+}
```
[the remark this answers is not in the document]
Comment (ealush) [reply packages/n4s/src/rules/schemaRules/tuple.ts:131]: This is consistent with how `shape`/`loose` work — they run each rule with the field value (or `undefined` for missing keys) and pass if the rule passes. There's no `__n4sOptional` marker system in enforce. A rule that accepts `undefined` (like `isNullish`) behaves identically in `shape` — if you have `enforce.shape({ x: enforce.isNullish() })`, then `enforce({}).shape(...)` passes because `isNullish().run(undefined)` succeeds.

Adding an explicit optional marker would be a cross-cutting refactor touching the optional rule infrastructure. Worth considering as a separate enhancement, but not a regression here since the behavior follows the existing pattern.

File: packages/n4s/src/rules/schemaRules/tuple.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+ * Used for the Args parameter of the returned RuleInstance.
+ */
+type InferTupleInput<T extends RuleInstance<any, any>[]> = {
+  [K in keyof T]: T[K] extends RuleInstance<any, [infer A, ...any[]]>
+    ? A
+    : never;
+};
+
+export type TupleRuleInstance<T extends RuleInstance<any, any>[]> =
+  RuleInstance<InferTuple<T>, [InferTupleInput<T>]>;
```
[the remark this answers is not in the document]
Comment (ealush) [reply packages/n4s/src/rules/schemaRules/tuple.ts:157]: Acknowledged — this is a known TypeScript limitation. Making trailing optionals reflected in the tuple type (e.g. `[string, number?]`) requires conditional type utilities like `TrimTrailingOptionals` and a type-level `IsOptionalRule` predicate. This adds significant type complexity for an edge case.

The tests use `@ts-expect-error` for these cases, and the runtime behavior is correct. Can be improved in a follow-up.
